src/context-menu/CanvasContextMenu.tsx

```diff
@@ -5,14 +5,83 @@ import { DiagramEngine, NodeModel, LinkModel } from '@projectstorm/react-diagram
 
 import '../../style/ContextMenu.css'
 import { commandIDs } from "../commands/CommandIDs";
-import { CustomPortModel } from '../components/port/CustomPortModel';
 
 export interface CanvasContextMenuProps {
 	app: JupyterFrontEnd;
 	engine: DiagramEngine;
 	nodePosition?: {x: number, y: number};
 }
 
+function customZoomToFit(
+    engine: DiagramEngine,
+    padding = 40
+) {
+    const model = engine.getModel();
+    const nodes = model.getNodes();
+    if (!nodes.length) return;
+
+    // 1) Reset
+    model.setZoomLevel(100);
+    model.setOffset(0, 0);
+
+    // 2) Compute model-space bounds
+    let minX = Infinity, minY = Infinity;
+    let maxX = -Infinity, maxY = -Infinity;
+    for (const node of nodes) {
+        const { x, y } = node.getPosition();
+        const { width, height } =
+        (node as any).getSize?.() ?? { width: 150, height: 100 };
+        minX = Math.min(minX, x);
+        minY = Math.min(minY, y);
+        maxX = Math.max(maxX, x + width);
+        maxY = Math.max(maxY, y + height);
+    }
+    // apply padding
+    minX -= padding;
+    minY -= padding;
+    maxX += padding;
+    maxY += padding;
+    const contentWidth  = maxX - minX;
+    const contentHeight = maxY - minY;
+
+    // 3) Measure viewport from Jupyter’s content widget
+    const contentWidget = document.querySelector<HTMLElement>(
+    '.lm-Widget[role="region"][aria-label="notebook content"]'
+    );
+    
+    let vpW: number, vpH: number;
+    const cs = window.getComputedStyle(contentWidget);
+    vpW = parseFloat(cs.width);
+    vpH = parseFloat(cs.height);
```

Comment:
instead of parsing pixel values from css, you can instead use `.getBoundingClientRect()` that will give you the width and height (along some other values).